### PR TITLE
Ensuring instruction labels are strings when provided to the Instruction constructor.

### LINKED ISSUE #7667: Submitting a non-string label to instructions leads to a TypeError when a QuantumCircuit instance is drawn.
### Environment

- **Qiskit Terra version**: 0.19.2
- **Python version**: 3.8.8
- **Operating system**: Pop!_OS 21.10

### What is happening?

While applying an `X` gate to two qubits, I made a mistake in which instead of passing in the target qubit indices as a list like so, `qc.x([0, 1])`, I submitted them as separate parameters as in,  `qc.x(0, 1)`. This lead to the following exception when I later called `qc.draw`:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/username/anaconda3/lib/python3.8/site-packages/qiskit/visualization/text.py", line 657, in __repr__
    return self.single_string()
  File "/home/username/anaconda3/lib/python3.8/site-packages/qiskit/visualization/text.py", line 665, in single_string
    return "\n".join(self.lines()).encode(self.encoding).decode(self.encoding)
  File "/home/username/anaconda3/lib/python3.8/site-packages/qiskit/visualization/text.py", line 709, in lines
    layers = self.build_layers()
  File "/home/username/anaconda3/lib/python3.8/site-packages/qiskit/visualization/text.py", line 1138, in build_layers
    layer, current_connections, connection_label = self._node_to_gate(node, layer)
  File "/home/username/anaconda3/lib/python3.8/site-packages/qiskit/visualization/text.py", line 1005, in _node_to_gate
    gate_text = gate_text + params
TypeError: unsupported operand type(s) for +: 'int' and 'str'
```
I quickly realized that I had failed to pass in the indices as a list, but was puzzled as to why I was getting this exception. I took a look at the [documentation](https://qiskit.org/documentation/stubs/qiskit.circuit.QuantumCircuit.x.html?highlight=x#qiskit.circuit.QuantumCircuit.x) corresponding to `x` and saw that the second index had been stored in the optional `label` parameter. This results in a `TypeError` being raised [here](https://github.com/Qiskit/qiskit-terra/blob/1ae663d80295dce1b6ff25cbf29c640bf2ff9952/qiskit/visualization/text.py#L1055) when the instruction is processed during a draw.

Note that a similar error arises, when one attempts to draw the circuit using Matplotlib, so this is not isolated to text based circuit displays:
```
Traceback (most recent call last):
  File "error.py", line 6, in <module>
    qc.draw(output='mpl')
  File "/home/username/anaconda3/lib/python3.8/site-packages/qiskit/circuit/quantumcircuit.py", line 1865, in draw
    return circuit_drawer(
  File "/home/username/anaconda3/lib/python3.8/site-packages/qiskit/visualization/circuit_visualization.py", line 284, in circuit_drawer
    image = _matplotlib_circuit_drawer(
  File "/home/username/anaconda3/lib/python3.8/site-packages/qiskit/visualization/circuit_visualization.py", line 678, in _matplotlib_circuit_drawer
    return qcd.draw(filename)
  File "/home/username/anaconda3/lib/python3.8/site-packages/qiskit/visualization/matplotlib.py", line 278, in draw
    self._get_layer_widths()
  File "/home/username/anaconda3/lib/python3.8/site-packages/qiskit/visualization/matplotlib.py", line 394, in _get_layer_widths
    and len(gate_text) < 3
TypeError: object of type 'int' has no len()
```

### How can we reproduce the issue?

This issue can be produced by executing the following code:
```
from qiskit import QuantumCircuit

qc = QuantumCircuit(2)
qc.x(0, 1)
print(qc)
```
Note that here I use a `print` instead of a `draw` to demonstrate that the same code is touched on either path.

### What should happen?

As a Qiskit user, I would expect an exception to be raised when I attempt to create an instruction while providing a label with an invalid type, rather than when I try to draw it. Currently, I receive an exception, but it's distanced from the root cause and isn't as clear or clean as a `ValueError` raised on instruction creation would be.

### Any suggestions?

I'm new to Qiskit, so forgive me if my suggestions are off base, but I did some sleuthing and determined a possible fix. Looking through the code, I found that all instructions have [Instruction](https://github.com/Qiskit/qiskit-terra/blob/1ae663d80295dce1b6ff25cbf29c640bf2ff9952/qiskit/circuit/instruction.py#L50) as their base class. If we were to add a type check to the [section](https://github.com/Qiskit/qiskit-terra/blob/1ae663d80295dce1b6ff25cbf29c640bf2ff9952/qiskit/circuit/instruction.py#L84) of `Instruction`'s constructor which sets the optional `label` when present, we could raise a `ValueError` that would notify the user of their mistake as soon as the instruction is created. I have a draft of such a fix working locally and would love to submit a pull request, if the group would like to assign this issue to me. Even if we don't go with my suggested fix, my willingness to help out remains.

### PR DESCRIPTION
**Pull request by upsideon**
### Summary

Fixes #7667 by adding a type check to the `Instruction` constructor which will raise a `TypeError` if the `label` parameter is provided, but not of type `str`. Per the associated issue, this guarantees that an informative exception is raised for labels of an invalid type at instruction creation time rather than when the label is later processed by `print` or `draw`.

### Details and comments

Note that this pull request also adds a test for the type check that already existed for the `label` setter, but which was not under test coverage. I thought it prudent to add this test so that if our type checking rules change for labels in the future, we immediately know via a test failure that a change needs to be applied in both locations.

### CODE REVIEW

File: test/python/circuit/test_instructions.py [hunk 1]
```diff
@@ -692,6 +692,19 @@ def dummy_requester(specifier):
             for instruction in instruction_list:
                 self.assertIs(instruction.condition[0], sentinel_register)
 
+    def test_label_type_enforcement(self):
+        """Test instruction label type enforcement."""
+        with self.subTest("accepts string labels"):
+            instruction = Instruction("h", 1, 0, [], label="label")
+            self.assertIs(instruction.label, "label")
```
Comment (jakelishman) [inline test/python/circuit/test_instructions.py:699]: Here you mean `self.assertEqual(instruction.label, "label")`.  `assertIs` is the same as `assert x is y` as opposed to `x == y`, but `is` is a referential check (i.e. "are these the same object in memory?") while `==` is an equality check ("do these objects have equal values?").  It's a CPython implementation detail that certain short string literals are cached in some circumstances, but we can't rely on it in general.

**Review by jakelishman — COMMENTED** [review]
Thanks for looking at this and making the PR - it looks good!  I left a couple of minor comments, but nothing serious.

Please could you also add an "upgrade" release note for this - the instructions should be in the [contributing guidelines](https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#release-notes).  We just need to explain in one sentence that instruction labels will now be type-checked on creation.

**Comment by upsideon** [comment]
Thanks for the feedback, @jakelishman! I've applied each of the requested changes.

**Comment by upsideon** [comment]
@jakelishman Do you think you'll have an opportunity to give this pull request another review?

**Review by jakelishman — APPROVED** [review]
Sorry for the delay!  This looks good to me now, and thanks for pinging me so it didn't get lost.

**Comment by jakelishman** [comment]
We're getting some test failures here after this change, but I actually think that this PR is turning up buggy behaviour from _existing_ code, and isn't the problem itself.  I'll have a look into it, and see what's going on there.

**Comment by upsideon** [comment]
Thank you for the additional review and thank you for looking into the test failures! Feel free to let me know if you'd like any help.

**Review by jakelishman — APPROVED** [review]
The issue turned out to be that a particular couple of automated tests for gates were guessing on how to construct everything, and in some cases it was getting it wrong and passing an extra parameter that was getting put in the `label` slot.  This PR then made that a failure (as it should be), so we saw a test failure.  I've pushed a commit that works around the issue (the automated instantiation is still pretty questionable), so this should be good to go now.  Thanks!

**Pull request merged**